### DISCUSSION TITLE
[Feature Request]: Option to limit traffic based on original hop settings

### DISCUSSION THREAD

**Discussion by PaulOckenden** [meshtastic/firmware#4060] [Ideas]
### Platform

other

### Description

Background: Here in the UK we have restricted airtime. And now in some parts of the country the mesh has become severely congested - almost to the point where it becomes unusable. This seems to have coincided with many users setting their max-hops to seven, rather than the recommended three.

Possible solution: An option in the Lora config, alongside the ability to set the hop count, which basically says "don't repeat anything with an original hop count higher than our local setting" - in other words just drop all these 7 hop packets.

The effect of this change would be twofold: Firstly it would start to reduce the congestion on the mesh. But secondly it would discourage people (especially those in areas with restricted airtime) from setting their hop count higher than the value that's strongly recommended in the Meshtastic docs! 

It would obviously need a firmware update and then additional work from those responsible for the clients, in order that this option could be enabled. Just needs a binary on/off via whatever UI device the client normally uses.

**Comment by BrianDrury** [comment]
Sounds like a very good idea to me.

**Comment by Jorropo** [comment]
A solution with less knobs could be to compute the next hop as `min(msg.current_hop -1, node.hop_limit)` so if you have MAX_HOP set to 3 and you receive a message with 6, you would relay with 3.

**Comment by PaulOckenden** [comment]
> A solution with less knobs could be to compute the next hop as `min(msg.current_hop -1, node.hop_limit)` so if you have MAX_HOP set to 3 and you receive a message with 6, you would relay with 3.

Possibly. But if big-hop packets were simply dropped and not forwarded by nodes using the option I suggested it might encourage people not to use mesh-unfriendly settings in the first place. Although I do understand that in a few very special circumstances using a large hop count might be beneficial. So in those areas people could simply leave the option unticked.

**Comment by Jorropo** [comment]
I see two issues with dropping them:
- it's hard to debug, you raise your hop limit and things starts not working because your neighbors configured it, you don't get any error message and it still kinda work, as zero-hop messages, so it's flaky and extremely hard to debug.
- it actually promote worst routing topologies, because you don't relay the message other nodes in the mesh that are closer and in worst position to relay the message might relay the message instead. Relaying with a lower number will make anyone else who heard the original 7 hop and your 3 hop relay, cancel their own relay attempt.

**Comment by noon92** [comment]
This is similar to #3833

**Comment by PaulOckenden** [comment]
> This is similar to #3833

Indeed. Although of course a repeater with ‘don’t decrypt’ enabled (as most routers seem to be) won’t do this. But I was thinking more of general nodes in the mesh, as opposed to routers and repeaters.

**Comment by GUVWAF** [comment]
@Jorropo Has two valid points as to why this would not be desirable. 

With the next-hop router (#2856), excess rebroadcasts due to high hop limits will be avoided for DMs already. Furthermore, already now the hop limit for ACKs and responses will be dynamically adapted based on the amount of hops that was required.

I'm converting this into a discussion. Don't think it's needed to implement this in the firmware. In my opinion you should either educate the people using your public mesh, or move to a private one. The hop limit is only one configuration setting that users can set wrongly/too aggressive. Enabling MQTT downlink or setting high update intervals for telemetry, position, etc. are also very important factors that contribute to the channel utilization.

**Reply by cwoollard** [reply]
Just a thought regarding what you have said. 

Would it be a good idea to have a Meshtastic Best Practices settings Page? That way new (& seasoned) Meshtastic users can see a complete list of recommended settings. Potentially, with a note saying (don't screw with this setting for x, y, z reason). Preferably in a table format.

**Reply by GUVWAF** [reply]
Yeah, that might be nice, but usually the "best practice" is just the default settings. That will work fine in a decent size mesh (20 nodes or so) and even larger if you're not duty cycle restricted. However, if you have a mesh of 100+ nodes with duty cycle, you have to turn down all periodic broadcasts and only increase hop limits on nodes on the edges.

**Reply by PaulOckenden** [reply]
> Yeah, that might be nice, but usually the "best practice" is just the default settings. That will work fine in a decent size mesh (20 nodes or so) and even larger if you're not duty cycle restricted. However, if you have a mesh of 100+ nodes with duty cycle, you have to turn down all periodic broadcasts and only increase hop limits on nodes on the edges.

It’d be great if the docs spelled that out. It’s all a bit vague. I’m in a duty cycle limited region, am in a mesh with hundreds of visible nodes, and more and more (currently nearing 50%) are set to 7 hops. If I watch the debug output on one of my nodes it’s just a constant stream of airtime restriction delays. So much traffic ends up getting dropped it’s killing Meshtastic. But the 7-hoppers think they are winning because they’ve collected loads of dots from hundreds of miles away. It doesn’t bother them that they can’t actually communicate with these nodes, it’s just a numbers game. Pokémon. Collecting nodes.

**Reply by GUVWAF** [reply]
"Really, 3 is fine": https://meshtastic.org/docs/configuration/radio/lora/#max-hops 
And in general: https://meshtastic.org/docs/configuration/tips/#best-practices

That said, setting hops to 7 will not lead to you hearing more nodes, but more nodes hearing you.

**Reply by PaulOckenden** [reply]
> "Really, 3 is fine": https://meshtastic.org/docs/configuration/radio/lora/#max-hops And in general: https://meshtastic.org/docs/configuration/tips/#best-practices
> 
> That said, setting hops to 7 will not lead to you hearing more nodes, but more nodes hearing you.

Again, I think that needs to be spelt out, otherwise we end up with the mess we currently have here. I've tried to politely ask people to calm their settings down and their response is that the docs only suggest 3 hops as a recommendation. One person suggested that if 7 hops wasn't be used then it shouldn't be in the settings. 

Perhaps the docs should also say that Meshtastic isn't Pokémon! ;-)

**Reply by GUVWAF** [reply]
There's a button "Edit this page" on each page of the website, feel free to add it anywhere you think it's still not clear.

**Comment by G0TJH-Ian** [comment]
It sounds just like what we want Paul

**Comment by andresgongora** [comment]
Writing for South Spain. We are running into the same situation you guys had in the UK 1 year ago. This is starting to become very popular here, but I'm experiencing issues with people abusing the hop count. Even though almost all nodes are >=2.5, I still experience usability issues.

I run many nodes in the area that I use to communicate with friends and family. I could configure them to only relay our own traffic, but that beats the purpose of contributing to a shared infrastructure.

The solution @Jorropo proposes sounds fair: "I'll relay messages, but not beyond my own hop count".

Is there anything like this in the works? I see this is a more generic case of #4764 .

**Reply by shalberd** [reply]
Agreed from a Swiss perspective. AI see @andresgongora 's pain point. Same pretty much across Europe, with only one frequency, 869.525 Mhz, regardless of which channel, which LoRa preset etc.
Yes, using a Lora preset with shorter TX times, like MEDIUM_FAST, can alleviate the situation a little, but the problem will pop up again eventually.
However, I also see the objection @Jorropo had regarding being careful not to introduce worse routing topologies / paths:

> Relaying with a lower number will make anyone else who heard the original 7 hop and your 3 hop relay, cancel their own relay attempt.

We have been somewhat successful at least convincing most people not to go beyond LoRa maxHops 5, but that is tough.
With no real intelligent routing layer in Meshtastic, I'd advocate for not allowing any maxHops greater than 5, but then again, I am just a humble Meshtastic fan from central Europe :-)

**Comment by Syridian** [comment]
From here in Canberra, AU, I’d love to see this implemented as an optional configuration item. It would save me from having to modify the firmware manually, and I think it could benefit many others in similar situations.

While there may be better ways to encourage responsible mesh behavior, I believe that visible configuration options can influence user choices. If users know that messages with high hop_count values might be ignored by some nodes, they may be less inclined to arbitrarily increase it just to force messages through. It’s a fine line between offering too many options and not enough — and I think this strikes a good balance.

There's always the question: Is this a social fix or a technological one? In some places, it's becoming clear that social fixes alone aren't working. A technological safeguard might feel harsh, but when it's in the hands of node operators, it becomes a form of community moderation — almost like voting with configuration.

That said, I do see the potential downside: nodes silently dropping high-hop messages could lead to poor node placement or even encourage users to find ways to circumvent the system, especially if dead spots emerge. So it’s important that this feature be opt-in, clearly documented, and used thoughtfully.

I’d love to hear others’ thoughts on whether this kind of filtering could help balance mesh health with user freedom — especially if implemented as a configurable option.

**Comment by dchokola** [comment]
A little disappointing that this is still a discussion after over two years.

The mesh is participatory. If you want to generate traffic, you must be willing to route it. I see it as analogous to a seed/leech ratio on torrents. In my area, I have all my nodes set up as clients with the default (3) hop limit. Unfortunately, many folks in the broader area have their hop limits set to 7. The conversations on the mesh are usually localized to 10s of miles, but I often get half-conversations on the default channel from 100s of miles away from my location, plus their telemetry. It's traffic my nodes are burdened with but which is of no benefit to the local mesh. Until I started manually adding nodes > 3 hops away to my ignore list, the *majority* (> 3/4) of the nodes in my nodelist were > 3 hops away.

It seems fair that all nodes should be required to route up to their own hop limit, and could opt in to rebroadcasting higher hops if they wanted. I consider it spam or leeching for folks to bump their hop limits high just to talk about the weather to their neighbors and I don't care if their routing is degraded because others refuse to rebroadcast it. I'll be running with custom firmware to enforce this.

**Comment by PaulOckenden** [comment]
Like you say, two years ago.

I got so fed up with meshtastic locally I've given up. Some others locally have moved to Meshcore.

I found it all fun for a couple of months. 

Interesting to see that people are still playing, though.

**Reply by shalberd** [reply]
Agreed, we had people here really trying to foster understanding and building custom firmware themselves. Disappointed because there is no way to configure such a threshold, they probably moved on to MeshCore https://github.com/meshtastic/firmware/discussions/4764#discussioncomment-10714821 https://github.com/meshtastic/firmware/discussions/4764#discussioncomment-10682868

I've got no idea whether the new traffic management feature in 2.8 will be a solution. In any case, it might be worth watching. However, I won't install unstable alpha firmware like 2.8 on my mountain nodes.

Traffic Management Module for packet forwarding — dedup, rate limiting, role-aware policing (https://github.com/meshtastic/firmware/pull/9358 base, https://github.com/meshtastic/firmware/pull/10706 dedup/rate-limit expansion, https://github.com/meshtastic/firmware/pull/10745 next-hop cache overflow store, https://github.com/meshtastic/firmware/pull/9921 congestion-aware position interval/hop-exhaustion tuning).
Automatic variable hop limits based on live mesh activity and message-size estimation (https://github.com/meshtastic/firmware/pull/10176).

**Comment by dchokola** [comment]
For 2.7.26, here's a quick hack which checks if any incoming packet has a `hop_start` greater than the node's own hop limit:
```
diff --git a/src/mesh/Router.cpp b/src/mesh/Router.cpp
index 96818339c..ad51bdf3b 100644
--- a/src/mesh/Router.cpp
+++ b/src/mesh/Router.cpp
@@ -725,6 +725,20 @@ NodeNum Router::getNodeNum()
  */
 void Router::handleReceived(meshtastic_MeshPacket *p, RxSource src)
 {
+    // ----------------------------------------------------------------------
+    // Custom Hop Limit Check: Drop packet if its max hop limit (hop_start)
+    // exceeds this node's configured max hop limit.
+    // ----------------------------------------------------------------------
+    uint8_t localMaxHopLimit = config.lora.hop_limit;
+    uint8_t packetStartHops = (p->hop_start > 0) ? p->hop_start : p->hop_limit;
+
+    if (packetStartHops > localMaxHopLimit) {
+        LOG_INFO("Dropping packet 0x%x: max hop limit (%d) exceeds local max hop limit (%d)",
+                 p->id, packetStartHops, localMaxHopLimit);
+        cancelSending(p->from, p->id);
+        return;
+    }
+
     bool skipHandle = false;
     // Also, we should set the time from the ISR and it should have msec level resolution
     p->rx_time = getValidTime(RTCQualityFromNet); // store the arrival timestamp for the phone
```
